### PR TITLE
Rework STM32 ADC sequencer

### LINKED ISSUE #61100: STM32U5 ADC: Selected channels are not reset between reads
**Describe the bug**
Performing an ADC read on a channel will leave that channel activated in the channel select register (e.g. ADC4->CHSEL). This will cause issues when attempting to read from a new channel, as the hardware will attempt to read from both.

**To Reproduce**
This can be seen using the adc driver demo. I modified the `nucleo_u575zi_q.overlay` to use ADC4 and include another channel:

```
/ {
	zephyr,user {
		/* adjust channel number according to pinmux in board.dts */
		io-channels = <&adc4 2>, <&adc4 3>;
	};
};

&adc4 {
	#address-cells = <1>;
	#size-cells = <0>;

	channel@2 {
		reg = <2>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		// Note the acquisiton time is set long to help demonstrate ISR firing multiple times for single request.
		zephyr,acquisition-time = <ADC_ACQ_TIME_MAX>;
		zephyr,resolution = <12>;
	};

	channel@3 {
		reg = <3>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_MAX>;
		zephyr,resolution = <12>;
	};
};
```

Then build the project with:

`west build --pristine -b nucleo_u575zi_q zephyr/samples/drivers/adc`

Running the project youll see the following:

1. first call to `adc_read` reading Channel 2: 
    * ADC4->CHSEL register == 0x04 (correct)
    * `adc_stm32_isr` is hit once for the single conversion.
2. second call to `adc_read` reading Channel 8: 
    * ADC4->CHSEL register == 0x0c (incorrect, should be 0x08)
    * `adc_stm32_isr`is hit twice, once for each channel. The destination buffer will overflow here from second read.

**Expected behavior**
A new read should not be impacted by any previous reads. Channels should be cleared before setting up the new read.

**Workaround:**
Reset the selected channels at the start of the `start_read` func in `adc_stm32.c`:

```c
static int start_read(const struct device *dev,
		      const struct adc_sequence *sequence)
{
	...

#if defined(CONFIG_SOC_SERIES_STM32U5X)
	// Note I've only tested with STM32U5 ADC4
	if (adc == ADC4) {
		LL_ADC_REG_SetSequencerChannels(adc, 0);
	}
#endif

	...
}
```

**Environment:**
 - OS: MacOS
 - Toolchain: Zephyr SDK, 
 - Zephyr v3.4

**Additional info:**
This issue might be present on other STM32 families as well. I dont any other family of boards to test, however a quick look over the `start_read` func doesnt seem to have any calls to reset the channels, and only calls to add new channels.

### PR DESCRIPTION
**Pull request by gautierg-st**
The ADC sequencer was not supported on all STM32 series.

This PR adds the support of the sequencer for all STM32 series with the exception of STM32F1 that needs DMA for that (and which is not fully supported yet).

Overlays of tests/drivers/adc_api have been updated in order to check the sequencer in automatic daily tests.

Fixes #61100

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -72,12 +72,43 @@ LOG_MODULE_REGISTER(adc_stm32);
...
-	!defined(CONFIG_SOC_SERIES_STM32F0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32G0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32WBAX) && \
-	!defined(CONFIG_SOC_SERIES_STM32WLX)
+#if defined(CONFIG_SOC_SERIES_STM32C0X) || \
+	defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X) || \
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:79]: Is it expected that U5 is both confirgurable and not configurable ?

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -72,12 +72,43 @@ LOG_MODULE_REGISTER(adc_stm32);
 /* reference voltage for the ADC */
 #define STM32_ADC_VREF_MV DT_INST_PROP(0, vref_mv)
 
-#if !defined(CONFIG_SOC_SERIES_STM32C0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32F0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32G0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
-	!defined(CONFIG_SOC_SERIES_STM32WBAX) && \
-	!defined(CONFIG_SOC_SERIES_STM32WLX)
+#if defined(CONFIG_SOC_SERIES_STM32C0X) || \
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:75]: A CONFIG_FULLY_CONFIGURABLE_SEQUENCE=y/n (set depending on series) would be much more elegant

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -823,6 +867,79 @@ static int set_resolution(const struct device *dev,
...
+			LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
+		}
+#endif /* FULLY_CONFIGURABLE_SEQUENCER */
+	}
+
+#if defined(NOT_FULLY_CONFIGURABLE_SEQUENCER)
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* On STM32U5 only ADC4 is not fully configurable */
+	if (adc == ADC4)
+#endif /* CONFIG_SOC_SERIES_STM32U5X */
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:916]: Looks like defining this in dt would simplify code a bit

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:79]: It is, I even put a comment about it below. ON U5, ADC1 and 2 are fully configurable, but not ADC4

File: drivers/adc/adc_stm32.c (hunk 2 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:75]: But for U5 it is both y and n...

File: drivers/adc/adc_stm32.c (hunk 3 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:916]: I could add a property for the type of sequencer, but some functions don't exist in some series, so I still need an #ifdef at some point.

Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:916]: Though I suppose I could create a macro similar to ANY_NUM_COMMON_SAMPLING_TIME_CHANNELS_IS to check if any ADC are either fully configurable or not to exclude the functions not present for the series.
I'll have a look at what is possible

**Comment by gautierg-st** [comment]
New version adds a new property for the type of sequencer used by the ADC.
A new macro  is used to not include functions that are not defined for the series.

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -823,6 +845,72 @@ static int set_resolution(const struct device *dev,
...
+			 */
+			LL_ADC_SetChannelPreselection(adc, channel);
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32U5X */
+			LL_ADC_REG_SetSequencerRanks(adc, table_rank[channel_index], channel);
+			LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
+		}
+#endif /* ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE) */
+	}
+
+#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:884]: Can we use `IS_ENABLED` here in order to gain readability ?
Same for all `#if ANY_ADC_SEQUENCER_TYPE_IS`

File: tests/drivers/adc/adc_api/testcase.yaml [hunk 5]
```diff
@@ -7,3 +7,10 @@ tests:
   drivers.adc:
     depends_on: adc
     min_flash: 40
+  drivers.adc.b_u585i_iot02a_adc4:
+    extra_args:
+      DTC_OVERLAY_FILE="boards/b_u585i_iot02a_adc4.overlay"
+    platform_allow:
+      - b_u585i_iot02a
+    integration_platforms:
+      - b_u585i_iot02a
```
Comment (erwango) [inline tests/drivers/adc/adc_api/testcase.yaml:16]: Is that required in this case ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:884]: You mean by replacing the ANY_ADC_SEQUENCER_TYPE_IS macro entirely? I don't know how to do that with just IS_ENABLED. The problem lies with U5 that has both a fully configurable ADC (ADC1/2) and a not fully configurable one (ADC4). In this case we need to include all the code. For all other series, we include only one of the two blocks of code.

Or do you mean something like this:
```
#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
#define IS_NOT_FULLY_CONFIGURABLE 1
#else
#define IS_NOT_FULLY_CONFIGURABLE 0
#endif
#if ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE)
#define IS_FULLY_CONFIGURABLE 1
#else
#define IS_FULLY_CONFIGURABLE 0
#endif

#if IS_ENABLED(IS_FULLY_CONFIGURABLE)
```
That's feasible, but I don't think it makes things more readable.

File: tests/drivers/adc/adc_api/testcase.yaml (hunk 5 above)
Comment (gautierg-st) [reply tests/drivers/adc/adc_api/testcase.yaml:16]: Probably not. I'll remove it

**Comment by gautierg-st** [comment]
Conflict fixed

**Review by erwango — DISMISSED** [review]

**Comment by kartben** [comment]
Doc failure is fixed (#62967) -- you may please rebase. Sorry for the inconvenience!

**Comment by gautierg-st** [comment]
> Doc failure is fixed (#62967) -- you may please rebase. Sorry for the inconvenience!

Rebase done. Thanks

**Review by FRASTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**